common/hyperscaler/rules/ruleset.go

```diff
@@ -0,0 +1,143 @@
+package rules
+
+import "fmt"
+
+type PatternAttribute struct {
+	matchAny bool
+	literal  string
+}
+
+type ValidRule struct {
+	Plan                    PatternAttribute
+	PlatformRegion          PatternAttribute
+	HyperscalerRegion       PatternAttribute
+	Shared                  bool
+	EuAccess                bool
+	PlatformRegionSuffix    bool
+	HyperscalerRegionSuffix bool
+	MatchAnyCount           int
+}
+
+type ValidationErrors struct {
+	ParsingErrors   []error
+	DuplicateErrors []error
+	AmbiguityErrors []error
+}
+
+func (pa *PatternAttribute) Match(value string) bool {
+	if pa.matchAny {
+		return true
+	}
+	return pa.literal == value
+}
+
+func (vr *ValidRule) Match(provisioningAttributes *ProvisioningAttributes) bool {
+	if !vr.Plan.Match(provisioningAttributes.Plan) {
+		return false
+	}
+
+	return vr.matchRegions(provisioningAttributes)
+}
+
+func (vr *ValidRule) matchRegions(provisioningAttributes *ProvisioningAttributes) bool {
```

Comment:
in general it is about matching input parameters, I know we have only regions as input params, but maybe the name could be about input params, nit regions